src/include/miopen/cat/problem_description.hpp

```diff
@@ -0,0 +1,146 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/problem_description_base.hpp>
+#include <miopen/activ.hpp>
+#include <miopen/tensor.hpp>
+
+#include <cassert>
+#include <string>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace cat {
+
+struct ProblemDescription : ProblemDescriptionBase
+{
+    ProblemDescription(const int32_t xCount_,
+                       const TensorDescriptor* const* xDescs_,
+                       const TensorDescriptor& yDesc_,
+                       int32_t dim_)
+        : xDescs(xDescs_), yDesc(yDesc_), xCount(xCount_), dim(dim_)
+    {
+    }
+
+    const TensorDescriptor& GetXDesc(int i) const
+    {
+        if(i >= xCount)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "CatForward: Invalid tensor index.");
+        }
+        return *xDescs[i];
+    }
+    const TensorDescriptor& GetYDesc() const { return yDesc; }
+    int32_t GetDim() const { return dim; }
+    int32_t GetXCount() const { return xCount; }
+
+    bool IsSameType() const
+    {
+        const auto dtype = yDesc.GetType();
+        for(int i = 0; i < xCount; i++)
+        {
+            if(xDescs[i]->GetType() != dtype)
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool IsRightLength() const
+    {
+        auto ydims = yDesc.GetLengths();
+        ydims[dim] = 0;
+        for(int i = 0; i < xCount; i++)
+        {
+            auto& xdims = xDescs[i]->GetLengths();
+
+            if(ydims.size() != xdims.size())
+            {
+                return false;
+            }
+
+            for(int j = 0; j < ydims.size(); j++)
+            {
+                if((j != dim) && (ydims[j] != xdims[j]))
+                {
+                    return false;
+                }
+            }
+            ydims[dim] += xdims[dim];
+        }
+
+        if(ydims[dim] != yDesc.GetLengths()[dim])
+        {
+            return false;
```

Comment:
@et16kr @Vsevolod1983 It was
```C++
MIOPEN_THROW(miopenStatusBadParm, "CatForward: Tensor dimension lengths do not match.");
```
which means that this `if` detects the case when PorblemConfig is incorrect.

❗ The proper approach is that ProblemConfig constructor must fail when/if the configuration is invalid/incorrect.

Knowing that, we can select one of the following approaches in the rest of the code:
- (A) ignore invalid configs (as we "assume" that invalid ProblemConfig can't be constructed
  - ⚠️ Assumptions are generally bad. Each assumption is an implicit dependence.
    - What if assumptions becomes incorrect in the future? When using assumptions we have to always analyze such scenarios.
- (B) detect and
  - (B.1) either throw exception
    - This is good for builds with `DEV=On` and for the _debug_ builds.
  - (B.2) or handle in some other way, if possible (like this function does now).
    - This approach is good for the _release_ builds.

Given that, the best approach for this function is to `return false` for the release builds and to `throw` for the debug/DEV builds. Please note that it is fine to ignore invalid ProblemConfigs in all other function of the Solver, because `IsApplicable()` is always called first. For example, we can safely ignore wrong ProblemConfigs in `GetSolution()`.

Action points:
- [ ] (1) Check if ProblemConfig constructor fails when/if the configuration is invalid/incorrect. Fix it, if necessary.
- [ ] (2) Consider if you would like to use the following approach:
```C++
#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
    MIOPEN_THROW(...); // this is (B.1)
#else
    return false;  // this is (B.2)
#endif
```

The final decision about AP (2) is totally up to you.

/FYI @CAHEK7 @DrizztDoUrden @averinevg @shurale-nkn

Reply:
I have modified the code as suggested.

Reply:
@et16kr Unresolved. This is not implemented yet:

> ❗ The constructor of ProblemConfig must fail when/if the configuration is invalid/incorrect.

When the above is implemented, MIOPEN_THROW can be removed from member functions because in this specific case it is guaranteed that the invalid ProblemConfig object won't be contructed.

Reply:
I have made changes to have the verification of dim and length in the constructor. 
Additionally, I removed exceptions from the IsAllPacked().